### PR TITLE
Changed Jersey implementation to OkHttp

### PR DESCRIPTION
**Pull request by lordofthejars**

### CODE REVIEW

**Comment by lordofthejars** [comment]
There are some things to note.

1) There is one failing test but it is also in current master branch.
2) I have added `https://github.com/SpectoLabs/hoverfly-java/pull/69/files#diff-50b108c74fa7a792d574d3093e70b66eR38` in some tests to restore system properties after test execution. If not done I got a lot of test failures due that system properties were not reset (look here https://github.com/SpectoLabs/hoverfly-java/blob/master/src/test/java/io/specto/hoverfly/junit/core/HoverflyTest.java#L250 system property is modified but never restored). What I am curious is why previous tests passed, but it seems (and I would like to know your opinion) is that Hoverfly-Java modifies some Java Properties but they don't restore back to original values after execution (or some of them).

If you want to test it you just need to comment the `Rule` in charge of restoring system environment and run tests. You'll see that failing tests are failing because `isHealth` method is pointing to a port not configured locally.

**Comment by lordofthejars** [comment]
I think that the best way to avoid any problem would be that Hoverfly-java takes care of storing original values of system properties and restore them when (stop) is called, then tests should not have to worry about system properties.

**Comment by tommysitu** [comment]
@lordofthejars Have you pushed all your changes? It is missing some dependencies.

**Comment by lordofthejars** [comment]
Yes. Which dependencies?

**Comment by tommysitu** [comment]
OkHttpClient and RestoreSystemProperties for JUnit. I can add them to build.gradle, but if you haven't done that already, then it's better for you to push your update.

**Comment by lordofthejars** [comment]
Done, it was my bad. I enabled logging in `build.gradle`, but then I run `checkout` command instead of just removing `testLogging`. Now everything is fixed.

File: src/test/java/io/specto/hoverfly/junit/core/HoverflyTest.java [hunk 1]
```diff
@@ -32,8 +35,12 @@
     private Hoverfly hoverfly;
     private ObjectMapper mapper = new ObjectMapper();
 
+    @Rule
+    public final RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();
```
Comment (tommysitu) [inline src/test/java/io/specto/hoverfly/junit/core/HoverflyTest.java]: You're right, Hoverfly should restore system properties in its own cleanup method. I will raise an issue for this bug.

File: src/main/java/io/specto/hoverfly/junit/dsl/RequestMatcherBuilder.java [hunk 2]
```diff
@@ -130,4 +128,25 @@ private String encodeUrl(String str) {
             throw new UnsupportedOperationException(e);
         }
     }
+
+    private static class MultivaluedHashMap<K, V> {
+        private Map<K, List<V>> elements = new HashMap<>();
+
+        public void add(K key, V value) {
```
Comment (tommysitu) [inline src/main/java/io/specto/hoverfly/junit/dsl/RequestMatcherBuilder.java]: Scope of the methods in this class can be package private

File: src/main/java/io/specto/hoverfly/junit/core/Hoverfly.java [hunk 3]
```diff
@@ -178,10 +176,24 @@ public void close() {
...
+            try {
+
+                final byte[] jsonContent = OBJECT_MAPPER.writeValueAsBytes(s);
+                RequestBody body = RequestBody.create(JSON, jsonContent);
+
+                final Request.Builder builder = createRequestBuilderWithUrl(SIMULATION_PATH);
+                final Request request = builder.put(body).build();
+
+                final Response response = client.newCall(request).execute();
+                if (!response.isSuccessful()) throw new IOException("Unexpected code " + response);
```
Comment (lordofthejars) [inline src/main/java/io/specto/hoverfly/junit/core/Hoverfly.java]: YEs this is something not clear to me, look this example from its website https://github.com/square/okhttp/wiki/Recipes#accessing-headers

Comment (lordofthejars) [inline src/main/java/io/specto/hoverfly/junit/core/Hoverfly.java]: The `toString` of Response logs everything, protocol, code, message and url.

Comment (tommysitu) [inline src/main/java/io/specto/hoverfly/junit/core/Hoverfly.java]: The Javadoc is more clear:  https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/ResponseBody.java
It suggests to use it in try-with resource block as well.

Comment (tommysitu) [inline src/main/java/io/specto/hoverfly/junit/core/Hoverfly.java]: Easy!

File: src/test/java/io/specto/hoverfly/junit/core/HoverflyTest.java (hunk 1 above)
Comment (bartoszmajsak) [inline src/test/java/io/specto/hoverfly/junit/core/HoverflyTest.java]: Are you guys started any work on that? If not I am happy to provide this enhancement in the PR as it's currently making my tests super ugly :) I have to deal with it on my own all over the place.

Comment (tommysitu) [inline src/test/java/io/specto/hoverfly/junit/core/HoverflyTest.java]: No I haven't. Feel free to work on this issue: https://github.com/SpectoLabs/hoverfly-java/issues/70
Thanks!

**Pull request merged**